# Data Wrangling

Before diving into the practical aspects of data wrangling, let's first understand the real meaning of the term "wrangle." In everyday language, it means to take care of or deal with someone or something, usually when it is challenging. Now, let's apply this concept to the world of data.

In the data world, "wrangling" is the process of dealing with difficult or messy data. This typically involves cleaning up raw, erroneous, or incomplete data, which can often consume 90% of a data scientist's or data analyst's time. The goal is to transform that messy data into a nice, structured, and easy-to-use format. This is the essence of data wrangling.

At this point, you might assume that data wrangling is simply another term for data cleaning, data remediation, or data munging. While there is overlap, data wrangling encompasses a broader set of processes.

These processes are designed to transform raw data into a format that is more readily usable for analysis. The specific steps involved can differ from project to project, depending on the nature of the data and the goals we are trying to achieve.


Some of the example include:

* Merging mulitple data sources into a single dataset for analysis
* Identifying gaps in data and either filling or deleting them
* Unncessary data removal which are irrelevant to the project
* Understanding of outliers(analysing the importance of outliers)

Note: Though data cleaning and data wrangling used interchangeably, there is distinction.

* Data wrangling is the overal process of transformation raw data into a more usable form.

* Data cleaning is a critical step in the data wrangling process to remove inaccurate and inconsistent data.

[Source](https://online.hbs.edu/blog/post/data-wrangling)

# Example: Used Cars Pricing

For this process I have used "Automobile Dataset" from the link:  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data


# Objectives

* Handling missing values
* Correct data formatting
* Standardize and normalize data

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

### Reading dataset from the URL

In [53]:
url =  'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'
df = pd.read_csv(url)
df

3    ?  alfa-romero     gas    std   two  convertible  rwd  front  88.60  \
0    3    ?  alfa-romero     gas    std   two  convertible  rwd  front   88.6   
1    1    ?  alfa-romero     gas    std   two    hatchback  rwd  front   94.5   
2    2  164         audi     gas    std  four        sedan  fwd  front   99.8   
3    2  164         audi     gas    std  four        sedan  4wd  front   99.4   
4    2    ?         audi     gas    std   two        sedan  fwd  front   99.8   
..  ..  ...          ...     ...    ...   ...          ...  ...    ...    ...   
199 -1   95        volvo     gas    std  four        sedan  rwd  front  109.1   
200 -1   95        volvo     gas  turbo  four        sedan  rwd  front  109.1   
201 -1   95        volvo     gas    std  four        sedan  rwd  front  109.1   
202 -1   95        volvo  diesel  turbo  four        sedan  rwd  front  109.1   
203 -1   95        volvo     gas  turbo  four        sedan  rwd  front  109.1   

     ...  130  mpfi  3.47  2.68  9.00  111  5000  21  27  13495  
0    ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
1    ...  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
2    ...  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
3    ...  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  
4    ...  136  mpfi  3.19  3.40   8.5  110  5500  19  25  15250  
..   ...  ...   ...   ...   ...   ...  ...   ...  ..  ..    ...  
199  ...  141  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
200  ...  141  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
201  ...  173  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
202  ...  145   idi  3.01  3.40  23.0  106  4800  26  27  22470  
203  ...  141  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[204 rows x 26 columns]

As there is no header in this dataset, so let's add the header columns.

In [3]:
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

In [4]:
df.columns = headers
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          1                 ?  alfa-romero       gas        std          two   
2          2               164         audi       gas        std         four   
3          2               164         audi       gas        std         four   
4          2                 ?         audi       gas        std          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1    hatchback          rwd           front        94.5  ...          152   
2        sedan          fwd           front        99.8  ...          109   
3        sedan          4wd           front        99.4  ...          136   
4        sedan          fwd           front        99.8  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  2.68    3.47               9.0        154      5000       19   
2         mpfi  3.19    3.40              10.0        102      5500       24   
3         mpfi  3.19    3.40               8.0        115      5500       18   
4         mpfi  3.19    3.40               8.5        110      5500       19   

  highway-mpg  price  
0          27  16500  
1          26  16500  
2          30  13950  
3          22  17450  
4          25  15250  

[5 rows x 26 columns]

We can clearly see the above there is '?'; Those may obviously may not reflect the essence in the data, which is gonna hinder furhter analysis.

__How to work with missing data?__

a. Identify missing data

b. Deal with missing data

c. Correct data format

# Identify and handle missing values

As above '?' has already identified pretty easily, so let's replace that with null values

In [5]:
# library
import numpy as np

# replace "?" to NaN

df.replace('?', np.nan, inplace = True)
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          1               NaN  alfa-romero       gas        std          two   
2          2               164         audi       gas        std         four   
3          2               164         audi       gas        std         four   
4          2               NaN         audi       gas        std          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1    hatchback          rwd           front        94.5  ...          152   
2        sedan          fwd           front        99.8  ...          109   
3        sedan          4wd           front        99.4  ...          136   
4        sedan          fwd           front        99.8  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  2.68    3.47               9.0        154      5000       19   
2         mpfi  3.19    3.40              10.0        102      5500       24   
3         mpfi  3.19    3.40               8.0        115      5500       18   
4         mpfi  3.19    3.40               8.5        110      5500       19   

  highway-mpg  price  
0          27  16500  
1          26  16500  
2          30  13950  
3          22  17450  
4          25  15250  

[5 rows x 26 columns]

### Evaluating for missing Data

There are two ways to find out the missing data.

a. isnull()

b. notnull()

In [6]:
df.isnull().head()

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False              False  False      False       False         False   
3      False              False  False      False       False         False   
4      False               True  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

In [26]:
df.notnull().head()

symboling  normalized-losses  make  fuel-type  aspiration  num-of-doors  \
0       True              False  True       True        True          True   
1       True              False  True       True        True          True   
2       True               True  True       True        True          True   
3       True               True  True       True        True          True   
4       True              False  True       True        True          True   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0        True          True             True        True  ...         True   
1        True          True             True        True  ...         True   
2        True          True             True        True  ...         True   
3        True          True             True        True  ...         True   
4        True          True             True        True  ...         True   

   fuel-system  bore  stroke  compression-ratio  horsepower  peak-rpm  \
0         True  True    True               True        True      True   
1         True  True    True               True        True      True   
2         True  True    True               True        True      True   
3         True  True    True               True        True      True   
4         True  True    True               True        True      True   

   city-mpg  highway-mpg  price  
0      True         True   True  
1      True         True   True  
2      True         True   True  
3      True         True   True  
4      True         True   True  

[5 rows x 26 columns]

The output is a boolean value indicating whether the value that is passed into the argument:

* is missing data in case of __isnull()__

* not missing data in case of __notnull()__


Let's visulise above boolean expression into another format using __sum()__ method, which returns the summation of total available missing values.

In [7]:
df.isnull().sum()

symboling             0
normalized-losses    40
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

We can also view above in anther way by using __value_counts()__ method.

In [8]:
# let's set above missing values in the new variable 'missing_data'

missing_data = df.isnull()

In [9]:
#for column in missing_data.columns.values.tolist():
#  print(column)
#  print("Missing data in", missing_data[column].value_counts())
#  print('')

### Deal with missing data:

1. Drop data:

  - Drop the whole row
  - Drop the whole column
2. Replace data

  - Replace with mean value
  - Replace by frequency
  - replace it based on other functions

Notes:

We should only drop the whole columns if most entries in the column are empty. In the data set, none of the columns are empty enough to drop entirely.

__Replace by mean:__

* "normalized-losses": 41 missind data, replace them with mean
* "stroke": 4 missing data, replace with mean
* "bore": 4 missind data, replace with mean
* "horsepower": 2 missing data, replace with mean
* "peak-rpm": 2 missing data, replace with mean


__Replace by Frequecny:__

* "num-of-doors": 2 missing data, replace with Frequency

__Drop the whole row:__

* "price": 4 missing data, simple delete the whole row

  - reason: For the prediction to price, we cannot use any data entry without price data, so better to remove them for now.




### Calculate the mean value for the 'normalized-losses' column

Before doing this, first we check the datatypes with __info() method. __

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 204 entries, 0 to 203
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          204 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               204 non-null    object 
 3   fuel-type          204 non-null    object 
 4   aspiration         204 non-null    object 
 5   num-of-doors       202 non-null    object 
 6   body-style         204 non-null    object 
 7   drive-wheels       204 non-null    object 
 8   engine-location    204 non-null    object 
 9   wheel-base         204 non-null    float64
 10  length             204 non-null    float64
 11  width              204 non-null    float64
 12  height             204 non-null    float64
 13  curb-weight        204 non-null    int64  
 14  engine-type        204 non-null    object 
 15  num-of-cylinders   204 non-null    object 
 16  engine-size        204 non

To calculate the mean value, string columns has to be converted into numeric type, which could be int or float depending upon the attributes.

For this we are coverting into float.

In [11]:
avg_norm_loss = df['normalized-losses'].astype('float').mean()
avg_norm_loss

np.float64(122.0)

### Replace "NaN with mean value

In [13]:
# replace on Normalised losses
df['normalized-losses'].replace(np.nan,avg_norm_loss, inplace = True)

/tmp/ipykernel_9504/2315536787.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['normalized-losses'].replace(np.nan,avg_norm_loss, inplace = True)


In [14]:
# replace on bore column
avg_bore = df['bore'].astype('float').mean()

df['bore'].replace(np.nan, avg_bore, inplace = True)

In [17]:
# replace on stroke column

avg_stroke = df['stroke'].astype('f').mean()

df['stroke'].replace(np.nan, avg_stroke, inplace = True)

In [19]:
#replace on horsepower

avg_horsepower = df['horsepower'].astype('f').mean()

df['horsepower'].replace(np.nan, avg_horsepower, inplace = True)

In [22]:
# replace on peakrpm
avg_peakrpm = df['peak-rpm'].astype('f').mean()
df['peak-rpm'].replace(np.nan, avg_peakrpm, inplace = True)

In [23]:
df['peak-rpm'].value_counts()

peak-rpm
5500           37
4800           36
5000           26
5200           23
5400           13
6000            9
4500            7
5800            7
5250            7
4200            5
4150            5
4750            4
4350            4
4250            3
5100            3
5900            3
4400            3
5125.990234     2
6600            2
4650            1
5600            1
5750            1
4900            1
5300            1
Name: count, dtype: int64

### Replace missing values by the most Frequent


How would we know now, by which number should we replace the nan values in case of categorical variables? 

* Check which categorical attributes has higher frequency values with two method: 

- __value_counts()__ : value_counts() method is used to see which values are present in a column. 
- __idxmax()__ : use to see the most common type or frequent one automatically without need to see any other values. Just quikly check which has got the higest count or common one. 
    * To use idxmax, add this "idxmax" method to the end of the 'value_counts()"

In [36]:
df['num-of-doors'].value_counts()

num-of-doors
four    116
two      88
Name: count, dtype: int64

* Four doors are the most common type. 

In [41]:
df['num-of-doors'].value_counts().idxmax()

'four'

In [42]:
# Replace with 'four'

df['num-of-doors'].replace(np.nan, 'four', inplace = True)

### Drop the rows entirely

In [48]:
# Drop whole row having NaN in Price column
df.dropna(subset=['price'], axis = 0, inplace= True)

__Note: We have to reset the index when rows are dropped (as in the previous step), the original row indices are retained. This method gives the DataFrame a new, sequential index starting from 0. There's gonna be old index as it is, with this argument, (drop = True) ensures that the old index (before resetting) is dropped.__

In [49]:
df.reset_index(drop= True, inplace= True)

In [52]:
#let's explore our data
df

symboling normalized-losses         make fuel-type aspiration  \
0            3             122.0  alfa-romero       gas        std   
1            1             122.0  alfa-romero       gas        std   
2            2               164         audi       gas        std   
3            2               164         audi       gas        std   
4            2             122.0         audi       gas        std   
..         ...               ...          ...       ...        ...   
195         -1                95        volvo       gas        std   
196         -1                95        volvo       gas      turbo   
197         -1                95        volvo       gas        std   
198         -1                95        volvo    diesel      turbo   
199         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two    hatchback          rwd           front        94.5  ...   
2           four        sedan          fwd           front        99.8  ...   
3           four        sedan          4wd           front        99.4  ...   
4            two        sedan          fwd           front        99.8  ...   
..           ...          ...          ...             ...         ...  ...   
195         four        sedan          rwd           front       109.1  ...   
196         four        sedan          rwd           front       109.1  ...   
197         four        sedan          rwd           front       109.1  ...   
198         four        sedan          rwd           front       109.1  ...   
199         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            152         mpfi  2.68    3.47               9.0        154   
2            109         mpfi  3.19    3.40              10.0        102   
3            136         mpfi  3.19    3.40               8.0        115   
4            136         mpfi  3.19    3.40               8.5        110   
..           ...          ...   ...     ...               ...        ...   
195          141         mpfi  3.78    3.15               9.5        114   
196          141         mpfi  3.78    3.15               8.7        160   
197          173         mpfi  3.58    2.87               8.8        134   
198          145          idi  3.01    3.40              23.0        106   
199          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  16500  
1        5000       19          26  16500  
2        5500       24          30  13950  
3        5500       18          22  17450  
4        5500       19          25  15250  
..        ...      ...         ...    ...  
195      5400       23          28  16845  
196      5300       19          25  19045  
197      5500       18          23  21485  
198      4800       26          27  22470  
199      5400       19          25  22625  

[200 rows x 26 columns]

Four rows has been deleted (200 rows) from the orignal data(204 rows)

In [55]:
# once again check the missing values
df.isna().sum()

3              0
?              0
alfa-romero    0
gas            0
std            0
two            0
convertible    0
rwd            0
front          0
88.60          0
168.80         0
64.10          0
48.80          0
2548           0
dohc           0
four           0
130            0
mpfi           0
3.47           0
2.68           0
9.00           0
111            0
5000           0
21             0
27             0
13495          0
dtype: int64

Our data looks with free of any missing values. 